### PR TITLE
 feat(transformer): strip implicit type export for typescript

### PR DESCRIPTION
**Pull request by magic-akari**

### CODE REVIEW

File: crates/oxc_transformer/src/typescript/mod.rs [hunk 1]
```diff
@@ -34,10 +37,21 @@ impl<'a> TypeScript<'a> {
...
+                    {
+                        decl.specifiers.retain(|specifier| {
+                            if specifier.export_kind.is_type() {
+                                return false;
+                            }
+
+                            let root_scope_id = self.ctx.scopes().root_scope_id();
+                            let name = specifier.local.name();
+
+                            self.ctx.scopes().has_binding(root_scope_id, name)
```
Comment (magic-akari) [inline crates/oxc_transformer/src/typescript/mod.rs:51]: How to distinguish whether a variable binding is a `Value` or a `Type` here?

File: crates/oxc_ast/src/syntax_directed_operations/bound_names.rs [hunk 2]
```diff
@@ -135,7 +135,9 @@ impl<'a> BoundNames for FormalParameter<'a> {
 impl<'a> BoundNames for ModuleDeclaration<'a> {
     fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
         if let ModuleDeclaration::ImportDeclaration(decl) = &self {
-            decl.bound_names(f);
+            if decl.import_kind.is_value() {
```
Comment (magic-akari) [inline crates/oxc_ast/src/syntax_directed_operations/bound_names.rs:138]: Which is better?
1. Skip the type only import binding
2. Add `TypeAlias` as includes SymbolFlags (but it's not `TypeAlias`, `ImporType` or something elese is better)

File: tasks/coverage/parser_babel.snap [hunk 3]
```diff
@@ -10004,16 +10005,6 @@ Expect to Parse: "typescript/types/const-type-parameters-babel-7/input.ts"
    ·       ╰── It can not be redeclared here
    ╰────
 
-  × Identifier `Context` has already been declared
```
Comment (magic-akari) [inline tasks/coverage/parser_babel.snap:10007]: We got the wrong result if we skip the type only import

File: crates/oxc_ast/src/syntax_directed_operations/bound_names.rs (hunk 2 above)
Comment (Boshen) [reply crates/oxc_ast/src/syntax_directed_operations/bound_names.rs:138]: `BoundNames` should align with the spec, so skip the type only import binding?

File: tasks/coverage/parser_babel.snap (hunk 3 above)
Comment (Boshen) [reply tasks/coverage/parser_babel.snap:10007]: It's fine for us to not report this typescript error, we are not a typescript checker after all ...

But it seems like we need to bind the symbols to the correct symbol flags at some point, but it's fine if this is skipped to get our transformer passed.

Comment (Boshen) [reply tasks/coverage/parser_babel.snap:10007]: By the way, symbol flag binding is directly copied from binder.ts, so you are free to modify the binder.

**Comment by Boshen** [comment]
This seems harder than I thought, I hope our infrastructure is fun for you to work with.

**Comment by Boshen** [comment]
Close or rebase?

**Comment by magic-akari** [comment]
I will rebase and try another way.

**Pull request closed without merging**